tokio/src/util/slab.rs

```diff
@@ -0,0 +1,745 @@
+
+            debug_assert_eq!(locked.slots.len(), locked.head);
+
+            Some((self.addr(idx), locked.slots[idx].gen_ref(self)))
+        }
+    }
+}
+
+impl<T> Page<T> {
+    /// Returns the slot index within the current page referenced by the given
+    /// address.
+    fn slot(&self, addr: Address) -> usize {
+        addr.0 - self.prev_len
+    }
+
+    /// Returns the address for the given slot
+    fn addr(&self, slot: usize) -> Address {
+        Address(slot + self.prev_len)
+    }
+}
+
+impl<T> Default for Page<T> {
+    fn default() -> Page<T> {
+        Page {
+            used: AtomicUsize::new(0),
+            slots: Mutex::new(Slots {
+                slots: Vec::new(),
+                head: 0,
+                used: 0,
+            }),
+            len: 0,
+            prev_len: 0,
+        }
+    }
+}
+
+impl<T> Page<T> {
+    /// Release a slot into the page's free list
+    fn release(&self, value: *const Value<T>) {
+        let mut locked = self.slots.lock().unwrap();
+
+        let idx = locked.index_for(value);
+        locked.slots[idx].next = locked.head as u32;
+        locked.head = idx;
+        locked.used -= 1;
+
+        self.used.store(locked.used, Release);
+    }
+}
+
+impl<T> CachedPage<T> {
+    /// Refresh the cache
+    fn refresh(&mut self, page: &Page<T>) {
+        let slots = page.slots.lock().unwrap();
+        self.slots = slots.slots.as_ptr();
+        self.init = slots.slots.len();
+    }
+
+    // Get a value by index
+    fn get(&self, idx: usize) -> &T {
+        assert!(idx < self.init);
+
+        // Safety: Pages are allocated concurrently, but are only ever
+        // **deallocated** by `Slab`. `Slab` will always have a more
+        // conservative view on the state of the slot array. Once `CachedPage`
+        // sees a slot pointer and initialized offset, it will remain valid
+        // until `compact()` is called. The `compact()` function also updates
+        // `CachedPage`.
+        unsafe {
+            let slot = self.slots.offset(idx as isize);
+            let value = slot as *const Value<T>;
+
+            &(*value).value
+        }
+    }
+}
+
+impl<T> Default for CachedPage<T> {
+    fn default() -> CachedPage<T> {
+        CachedPage {
+            slots: ptr::null(),
+            init: 0,
+        }
+    }
+}
+
+impl<T> Slots<T> {
+    /// Maps a slot pointer to an offset within the current page.
+    ///
+    /// # Panics
+    ///
+    /// panics if the provided slot pointer is not contained by the page.
+    fn index_for(&self, slot: *const Value<T>) -> usize {
+        use std::mem;
+
+        let base = &self.slots[0] as *const _ as usize;
+
+        assert!(base != 0, "page is unallocated");
+
+        let slot = slot as usize;
+        let width = mem::size_of::<Slot<T>>();
+
+        assert!(slot >= base, "unexpected pointer");
+
+        let idx = (slot - base) / width;
+        assert!(idx < self.slots.len() as usize);
+
+        idx
+    }
```

Comment:
I suppose there may be enough I/O resources that shaving off an extra usize is worth doing a big pile of pointer math?

Reply:
That was the reasoning. I can add a comment.